### DISCUSSION TITLE
How do `flip` and `shift` boundaries work?

### DISCUSSION THREAD

**Discussion by eternalthinker** [floating-ui/floating-ui#2916] [Q&A] [answered]
Hi all, I am trying to understand how the boundary argument works for the `flip` (and `shift`) middleware in Floating UI.

As I understand, the default value for boundary is `clippingAncestors`. But, when the reference and floating element are inside a scrollable container, the flip and shift don't respect the bounds of the scrolling container (Not flipping or shifting). They seem to be considering only the rootBoundary (`viewport`?). To make this work, I have to set the boundary as the scrolling container div explicitly.
In contrast, the `hide` middleware triggers referenceHidden when it scrolls past the bounds of the scrolling container.

See a codepen here: https://codepen.io/eternalthinkerr/pen/XWwdMZE
For scroll parent boundary option, I am explicitly setting the boundary. The default option triggers flip and shift only at the viewport.

Is this expected behaviour? Why does `clippingAncestors` boundary value for flip ignore the scrolling container?

**Comment by atomiks** [comment, marked as answer]
`clippingAncestors` as the name implies only finds ancestors that will clip the floating element. Since the scrolling container doesn't clip the floating element, it's not considered a required boundary.

Other than passing an explicit container, specifying `elementContext: 'reference'` in `flip`/`shift` detects `clippingAncestors` to be the same as what the reference's one is. This lets the floating element flip and shift if part of the reference will be clipped, before it will appear partially detached. However, it can still escape the scrolling container boundary partially in this scenario, unlike an explicit boundary - so it depends what you want.

**Reply by eternalthinker** [reply]
Thanks for the clarification @atomiks!
Yes, using the reference elementContext isn't a neat solution as it doesn't properly contain the floating element within the scrolling bounds.

For context, I am working on migrating from the older Popper v1 to Floating UI.
I have some components that use Popper's `scrollParent` boundary, which has the nice benefit of containing the floating content within the scrolling container, without explicitly knowing what the container element is.
Is my understanding correct that we no more have a way to do this implicitly in Floating UI?

**Reply by atomiks** [reply]
There is no string option for it. You'll need to make the scrolling container a `clippingAncestor` of the floating element as well. This can be done by making the scrollable container a containing block and rendering the floating element inside of it.

```js
<div style="will-change: transform; overflow: scroll;">
  <button>trigger</button>
  <div>tooltip</div>
</div>
```

Props that create a containing block:

- `transform`
- `will-change`
- `filter` etc.

---

Also we removed this in Popper v2 in 2020, since it was a poor default, so it hasn't been around for a while now. The reason is arbitrary 0-height `overflow: hidden` containers could cause the floating element to be incorrectly positioned as it tries to avoid something it doesn't need to. This was very common and for generic solutions is not a good default.

Usually the floating element can appear on the top layer over everything, avoiding scrolling containers and such rather than flipping inside them. Since this has rarely been brought up since, it's not really an important use case and maybe you could reconsider actually needing/using it? (I think `elementContext: 'reference'` is actually better overall.)

**Reply by eternalthinker** [reply]
I agree that the scroll parent is not a great default. In fact, we have been defaulting to viewport for the majority of usages.
However, there are specific cases where the elements and their floating labels need to be contained in a 'world' within the scroll boundaries, and where overflowing floating content will look awkward. Will look into a custom solution for these.
Thanks for the detailed explanation and hints!
